File: Easy/888 - FairCandySwap/solution.ts
// Question 888. Fair Candy Swap (https://leetcode-cn.com/problems/fair-candy-swap/)

/**
 * 执行用时：120 ms, 在所有 TypeScript 提交中击败了 78.29% 的用户
 * 内存消耗：49.2 MB, 在所有 TypeScript 提交中击败了 20.39% 的用户
 *
 * @param {number[]} A
 * @param {number[]} B
 * @returns {number[]}
 */
function fairCandySwap(A: number[], B: number[]): number[] {
  let result: number[] = [];

  const aCandy: Map<number, number> = new Map<number, number>();
  for (const a of A) {
    aCandy.set(a, aCandy.has(a) ? aCandy.get(a) + 1 : 1);
  }

  const halfBAGap: number = (getSum(B) - getSum(A)) / 2;
  for (const b of B) {
    const x = b - halfBAGap;
    if (aCandy.has(x)) {
      return [x, b];
    }
  }

  return result;
}

function getSum(array: number[]): number {
  return array.reduce((acc, current) => acc + current, 0);
}

File: Easy/888 - FairCandySwap/Solution.cs
// Question 888. Fair Candy Swap (https://leetcode-cn.com/problems/fair-candy-swap/)

/// <summary>
/// 执行用时：376 ms, 在所有 C# 提交中击败了 86.09% 的用户
/// 内存消耗：47.6 MB, 在所有 C# 提交中击败了 37.96% 的用户
/// </summary>
public class Solution 
{
    public int[] FairCandySwap(int[] A, int[] B) 
    {
        int[] answer = new int[2];

        Dictionary<int, int> aCandy = new Dictionary<int, int>();
        int aCandyCount = A.Count();
        for (var i = 0; i < aCandyCount; i++)
        {
            int currentValue = A[i];
            aCandy[currentValue] = aCandy.ContainsKey(currentValue) ? aCandy[currentValue] + 1 : 1;
        }

        int aSum = A.Sum();
        int bSum = B.Sum();

        int halfBAGap = (bSum - aSum) / 2;

        int bCandyCount = B.Count();
        for (int j = 0; j < bCandyCount; j++)
        {
            int y = B[j];
            int x = y - halfBAGap;
            if (aCandy.ContainsKey(x))
            {
                answer[0] = x;
                answer[1] = y;
                break;
            }
        }
    

        return answer;
    }
}
